[AIRFLOW-6541] Use EmrJobFlowSensor for other states

mustafagok opened the pull request:
Currently, EmrJobFlowSensor waits cluster to be terminated (state: "TERMINATED"). It can be used for other purposes if cluster states to be checked (i.e., NON_TERMINAL_STATES) can be passed to the operator.

For example, this sensor could be used to wait cluster to be ready (state: "RUNNING" or "WAITING").

Remove class constants of EmrBaseSensor: NON_TERMINAL_STATES, FAILED_STATE. Add new parameters: target_states and failed_states. Eliminate pylint warnings. Do not change default behaviour of the updated sensors.

---
Issue link: [AIRFLOW-6541](https://issues.apache.org/jira/browse/AIRFLOW-6541)

- [x] Description above provides context of the change
- [x] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

turbaszek commented on airflow/contrib/sensors/emr_job_flow_sensor.py:
```diff
@@ -16,48 +16,96 @@
         super().__init__(*args, **kwargs)
         self.job_flow_id = job_flow_id
+        if target_states is None:
+            target_states = ['TERMINATED']
```
What about oneliner?
```suggestion
            target_states = target_states or ['TERMINATED']
```

And shouldn't we extend the list to `['STARTING', 'BOOTSTRAPPING', 'RUNNING', 'WAITING', 'TERMINATING']` to preserve backward compatibility?

mustafagok replied:
Actually I have reversed the state logic, previously it was checking whether the state is in the `NON_TERMINAL_STATES` or not, if it is in the list, it was returning `False`. Now it is checking whether the state is in the `self.target_states` or not, if it is in it returns `True`. I could say `self.non_target_states` but is does not seem well.
I am sure that it is bacward compatible with this list, and if it was not, the unit test would fail. (I have also updated them to simplify, there were 4 classes almost same, I changed them with 1 common class).

oneliner is good idea, I will do it.

turbaszek commented on airflow/contrib/sensors/emr_job_flow_sensor.py:
```diff
@@ -16,48 +16,96 @@
+            failed_states = ['TERMINATED_WITH_ERRORS']
+        self.failed_states = failed_states
 
     def get_emr_response(self):
```
```suggestion
    def get_emr_response(self) -> Dict[str, Any]:
```

turbaszek commented on airflow/contrib/sensors/emr_job_flow_sensor.py:
```diff
@@ -16,48 +16,96 @@
         return emr.describe_cluster(ClusterId=self.job_flow_id)
 
     @staticmethod
     def state_from_response(response):
```
```suggestion
    def state_from_response(response: Dict[str, Any]) -> str:
```
It's awesome that you've added the docstrings! Type annotations will help even more.

turbaszek commented on airflow/contrib/sensors/emr_step_sensor.py:
```diff
@@ -16,52 +16,102 @@
+
     :param job_flow_id: job_flow_id which contains the step check the state of
     :type job_flow_id: str
+
```
We rather do not use new line here

mustafagok replied:
Some of the classes has this, and it seems better to me. But if you want so, I will remove all of the empty line between param-type groups.

turbaszek reviewed (commented):
Nice work! I have a few small comments.

mustafagok committed:
[AIRFLOW-6541] Use EmrJobFlowSensor for other states

Remove class constants of EmrBaseSensor: NON_TERMINAL_STATES,
FAILED_STATE. Add new parameters: target_states and failed_states.
Eliminate pylint warnings. Do not change default behaviour of updated
sensors.

turbaszek reviewed (approved):
LGTM!
